src/ruby/puppetserver-lib/puppet/server/master.rb

```diff
@@ -127,6 +162,79 @@ def getTaskData(environment_name, module_name, task_name)
 
   private
 
+  # @return Puppet::Node::Facts facts, Hash trusted_facts
+  def process_facts(request_data)
+    if request_data['facts'].nil?
+      facts = get_facts_from_pdb(request_data['certname'], request_data['environment'])
+    else
+      facts_from_request = request_data['facts']
+      facts_from_request['name'] = request_data['certname']
+      facts = Puppet::Node::Facts.from_data_hash(facts_from_request)
+    end
+
+    fact_values = if facts.nil?
+                    {}
+                  else
+                    facts.sanitize
+                    facts.to_data_hash['values']
+                  end
+
+    # Pull the trusted facts from the request, or attempt to extract them from
+    # the facts hash
+    trusted_facts = if request_data['trusted_facts'].nil?
+                      fact_values['trusted'].nil? ? {} : fact_values['trusted']
+                    else
+                      request_data['trusted_facts']['values']
+                    end
+
+    return facts, trusted_facts
+  end
+
+  def get_facts_from_pdb(nodename, environment)
+    if Puppet[:storeconfigs_backend] == :puppetdb
```

Comment:
I think this will always be `:puppetdb` regardless if PuppetDB is actually installed or configured - [storeconfig_backend](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/defaults.rb#L1891-L1898).

We might need to check both the backend and `Puppet[:storeconfigs] == true`? Or should we call `if Puppet::Node::Facts.indirection.class == :puppetdb`?

It looks like PuppetDB encourages consumers to configure the fact cache class to be `:yaml` and the generic fact indirection to be `:puppetdb` - [PuppetDB configuration](https://puppet.com/docs/puppetdb/6.2/connect_puppet_master.html#edit-routesyaml)??

I'm a little unclear how that corresponds to setting the facts cache class in the [`:storeconfigs` setting](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/defaults.rb#L1886)?

Reply:
Oh good point about the default. I switched to using that after setting up PDB for myself (the instructions there say to set that setting to `puppetdb`), because the problem with using `Puppet::Node::Facts.indirection.class == :puppetdb` is that it still relies on the global configuration of the indirector, which I felt kind of defeated the point. I think maybe checking `Puppet[:storeconfigs] == true` makes more sense here.

The connection between the hook and the actual configured terminii seems to be https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/store_configs.rb#L6 for the terminus class and https://github.com/puppetlabs/puppet/blob/1a63db161de417c9dcbeffeb5b60790a6aee6503/lib/puppet/application_support.rb#L57 for the cache class. However, If we instantiate the terminus ourselves, we lose the automatic caching configured by `routes.yaml`.

Reply:
For now I have switched this to use `Puppet['storeconfigs'] == true`, later we can switch over to using Justin's cached indirector data if we like that idea, see https://gist.github.com/justinstoller/db1b9b761aa43e55615d787a49b7f9e7.